spec/bar-chart-spec.js

```diff
@@ -1278,6 +1278,81 @@ describe('dc.BarChart', () => {
         }
     });
 
+    describe('accessibility bar chart', () => {
+
+        function removeEmptyBins (sourceGroup) {
+            return {
+                all:function () {
+                    return sourceGroup.all().filter( d => d.value !== 0);
+                }
+            };
+        }
+
+        it('internal elements are focusable by keyboard', () => {
+            chart.keyboardAccessible(true);
+            chart.render();
+            chart.selectAll('rect.bar').each(function () {
+                const bar = d3.select(this);
+                expect(bar.attr('tabindex')).toEqual('0');
+            });
+        });
+
+        it('initial internal elements are clickable by pressing enter', () => {
+            chart.keyboardAccessible(true);
+            const clickHandlerSpy = jasmine.createSpy();
+            chart.onClick = clickHandlerSpy;
+            chart.render();
+          
+            const event = new Event('keydown');
+            event.keyCode = 13;
+                     
+            chart.selectAll('rect.bar').each(function () {
+                const bar = d3.select(this).node();
+                bar.dispatchEvent(event);
+                expect(clickHandlerSpy).toHaveBeenCalled();            
+            });
+        });
+
+        it('newly added internal elements are also clickable from keyboard', () => {
+            chart.keyboardAccessible(true);
+            const clickHandlerSpy = jasmine.createSpy();
+            chart.onClick = clickHandlerSpy;
+
+            const event = new Event('keydown');
+            event.keyCode = 13;
+
+            const stateDimension = data.dimension(d => d.state);
+            const regionDimension = data.dimension(d => d.region);
+            const stateGroup = stateDimension.group().reduceSum(d => +d.value);
+            const ordinalDomainValues = ['California', 'Colorado', 'Delaware', 'Ontario', 'Mississippi', 'Oklahoma'];
+            
+            chart
+                .dimension(stateDimension)
+                .group(removeEmptyBins(stateGroup))
+                .xUnits(dc.units.ordinal)
+                .x(d3.scaleBand().domain(ordinalDomainValues))
+                .elasticX(true)
+                .barPadding(0)
+                .outerPadding(0.1)
+                .transitionDuration(0)
+                .render();
+
+            //force creation of new <rect> elements on existing chart
+            regionDimension.filterExact('West');
+            chart.redraw();
+            regionDimension.filterAll();
+            chart.redraw();
+                    
+            chart.selectAll('rect.bar').each(function () {
+                const bar = d3.select(this).node();
+                bar.dispatchEvent(event);
+                expect(clickHandlerSpy).toHaveBeenCalled();            
+            });
```

Comment:
```suggestion
            chart.selectAll('rect.bar').each(function (d) {
                this.dispatchEvent(event);
                expect(clickHandlerSpy).toHaveBeenCalledWith(d);
            });

```
1. It's better to check that the correct thing has been clicked, rather than just checking if the spy has been called. Also,  `toHaveBeenCalled` will always be true after the first iteration, because the spy is keeping a count of all calls since it was created.
2. `const bar = d3.select(this).node();` is the same as saying `const bar = this;`, and the fact that `this` is the bar should be apparent from the `selectAll` above.

Reply:
Hmmm, thinking about this further, `toHaveBeenCalledWith` will check all calls ever. We 

want to see that the last call equals the datum, so I guess we need 

```js
expect(obj.method.calls.mostRecent().args[0]).toEqual(d)
```
to be precise.

Reply:
I got it to work like this

```suggestion
            const clickHandlerSpy = jasmine.createSpy();
            chart.onClick = clickHandlerSpy;
            chart.selectAll('rect.bar').each(function (d) {
                this.dispatchEvent(event);
                expect(clickHandlerSpy).toHaveBeenCalledWith(d);
                clickHandlerSpy.calls.reset();
            });
```

Seems excessively complicated, but this clearly checks that the particular event caused the particular datum to get selected.

Reply:
Thanks, that's good to know. I'll apply your fix to all the other tests that check this functionality.

Reply:
Great, thanks!

Reply:
Done!